### PR TITLE
genesets e2e tests

### LINKED ISSUE #2042: Gene sets end to end tests (eg, puppeteer tests)

### PR DESCRIPTION
**Pull request by colinmegill**
Closes #2042

### CODE REVIEW

File: client/Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 include ../common.mk
 
-ANNOTATIONS := $(if $(ANNOTATIONS),$(ANNOTATIONS),../backend/test/fixtures/pbmc3k-annotations.csv)
+TEST_CSVS := $(if $(ANNOTATIONS),$(ANNOTATIONS),../backend/test/fixtures/{pbmc3k-annotations.csv, pbmc3k-genesets.csv})
```
Comment (seve) [inline client/Makefile:3]: try 
```suggestion
TEST_CSVS := $(if $(ANNOTATIONS),$(ANNOTATIONS),../backend/test/fixtures/{pbmc3k-annotations.csv,pbmc3k-genesets.csv})
```

File: client/__tests__/e2e/e2eAnnotations.test.js [hunk 2]
```diff
@@ -48,6 +84,87 @@ async function setup(config) {
   await waitByClass("autosave-complete");
 }
 
+describe("geneSET crud operations and interactions", () => {
+  test("static pre existing genesets load from csv, expand", async () => {
```
Comment (seve) [inline client/__tests__/e2e/e2eAnnotations.test.js:88]: There are no static gene sets and the test doesn't attempt to expand here

**Review by seve — APPROVED** [review]

File: backend/test/fixtures/pbmc3k-genesets.csv [hunk 3]
```diff
@@ -1,16 +1,21 @@
...
 fifth_dataset,,,
 summary test,,ACD,
 summary test,,AATF,
 summary test,,F5,
 summary test,,PIGU,
+geneset_to_delete,,,
+geneset_to_edit,,,
+fill_this_geneset,,RER1,
+empty_this_geneset,,SIK1,
+brush_this_gene,,SIK1,
```
Comment (bkmartinjr) [inline backend/test/fixtures/pbmc3k-genesets.csv:21]: missing EOL in this file

File: .gitignore [hunk 4]
```diff
@@ -58,3 +58,7 @@ client/.eslintcache
 
 # E2E Testing
 ignoreE2E*
+test.csv
+test_genesets.csv
+test-backups/
+test_genesets-backups/
```
Comment (bkmartinjr) [inline .gitignore:64]: I believe these names are unique to Colin's personal workflow.  Why add to the repo?

File: client/__tests__/e2e/e2eAnnotations.test.js [hunk 5]
```diff
@@ -48,6 +84,87 @@ async function setup(config) {
...
+  test("create a new geneset", async () => {
+    await goToPage(appUrlBase);
+
+    const genesetName = `test-geneset-foo-123`;
+    await assertGenesetDoesNotExist(genesetName);
+    await createGeneset(genesetName);
+    /* note: as of June 2021, the aria label is in the truncate component which clones the element */
+    await assertGenesetExists(genesetName);
+  });
+  test("edit geneset name", async () => {
```
Comment (bkmartinjr) [inline client/__tests__/e2e/e2eAnnotations.test.js:122]: optional: you might consider linting this code, as it doesn't match our white space rules.

File: client/__tests__/e2e/e2eAnnotations.test.js [hunk 6]
```diff
@@ -48,6 +84,87 @@ async function setup(config) {
...
+    await assertGenesetExists(newGenesetName);
+  });
+  test("delete a geneset", async () => {
+    await goToPage(appUrlBase);
+
+    await deleteGeneset(genesetToDeleteName);
+  });
+});
+
+describe("GENE crud operations and interactions", () => {
```
Comment (bkmartinjr) [inline client/__tests__/e2e/e2eAnnotations.test.js:135]: I like that the tests focus on the happy path primarily.  But there are some very important happy path tests that you need to add, as they are historically where we see bugs.  I suggest:
* color-by gene set & gene in a gene set
* test on subset, not just whole (there is already a framework within this test file to do so)
* test with clip
* undo/redo when you create, edit or delete a geneset and/or a gene in a geneset

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
I would like to see a more complete set of tests for the happy path, in particular those areas that have complex interaction with other parts of the system (color-by, subset, clip, undo/redo).   There is support for this already in the test framework, and lots of examples in the e2e annotations test file.

File: .gitignore [hunk 7]
```diff
@@ -57,4 +57,4 @@ docs/Gemfile.lock
 client/.eslintcache
 
 # E2E Testing
-ignoreE2E*
+ignoreE2E*
```
Comment (bkmartinjr) [inline .gitignore:60]: line is missing an EOL

File: backend/test/fixtures/pbmc3k-genesets.csv [hunk 8]
```diff
@@ -1,16 +1,21 @@
 # Test fixture
 gene_set_name,gene_set_description,gene_symbol,gene_description
-first gene set name,,F5, a gene_description
+first gene set name,a description,F5, a gene_description
```
Comment (bkmartinjr) [inline backend/test/fixtures/pbmc3k-genesets.csv:3]: why did you change the descriptions on line 3, 6, and 7?  This changes the test (ie, it no longer validates that we correctly merge different descriptions for a gene set). If you have tests that require this change, please add a new gene set to test it on.

File: backend/test/fixtures/pbmc3k-genesets.csv [hunk 9]
```diff
@@ -1,16 +1,21 @@
...
 first gene set name,a description,NO_SUCH_GENE, non-existent gene
 first gene set name,a description,F5, duplicate gene
-first gene set name, a description,SUMO3,
-first gene set name,,SRM,
-second gene set,,RER1
-second gene set,,SIK1
-third gene set,,NO_SUCH_GENE
+first gene set name,a description,SUMO3,
+first gene set name,a description,SRM,
+second_gene_set,,RER1,
```
Comment (bkmartinjr) [inline backend/test/fixtures/pbmc3k-genesets.csv:8]: line 8 and 9 were valid (ie missing a comma) and were there to test that we correctly handle it.  What was the reason for this change?

File: client/src/components/geneExpression/geneSet.js [hunk 10]
```diff
@@ -81,25 +81,33 @@ class GeneSet extends React.Component {
...
   }
 
   render() {
     const { setName, setGenes, genesetDescription } = this.props;
     const { isOpen } = this.state;
     const genesetNameLengthVisible = 150; /* this magic number determines how much of a long geneset name we see */
     const genesetIsEmpty = setGenes.length === 0;
+    let testClass = "geneset-expand";
+
+    if (setName.includes("Pop1 high")) testClass = "pop-1-geneset-expand";
```
Comment (bkmartinjr) [inline client/src/components/geneExpression/geneSet.js:109]: this is very, very fragile.  If you change the name in the reducer, it will break the tests.  Also, the reducer really should not contain the UI (ie, the population names) - they should be in the component.

I think this information should be passed as flags, not inferred from UI/presentation strings.

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
Punch list of requests:
* please fix PR description
* resolve comments if they are handled
* typo in .gitignore
* the changes to pbmc3k-genesets.csv (test fixture) invalidate some of the unit tests.  Was that intentional?  If not, please revert.
* the method used to detect the diffexp gene sets in geneExpression/geneSet.js is super fragile.  Please use a flag or derive it from API state, not UI strings.

**Review by bkmartinjr — APPROVED** [review]
LGTM. Thanks for the rapid turn-around on the punch list!
